### PR TITLE
Fix some file mode bits missing when doing mount syscall

### LINKED ISSUE #3952: [v1.1.8 regression] Sticky bit on container tmpfs no longer set
### Description

On runc 1.1.7 our containers had a tmpfs mounted in, which as expected had the sticky bit set:

```
app@d28c3e42b1a4:~$ ls -lah / | grep tmp
drwxrwxrwt   2 root root   40 Aug  1 05:09 tmp
```

Upon upgrading to runc 1.1.8 (via containerd 1.6.22) it seems that the sticky bit is no longer set on the tmpfs:

```
app@3aab0377e072:~$ ls -alh / | grep tmp
drwxrwxrwx   2 root root   40 Aug  1 04:43 tmp
```

At a glance, this seems to be a regression introduced in https://github.com/opencontainers/runc/pull/3916

### Steps to reproduce the issue

1. Upgrade to runc 1.1.8
2. Run a container with the following configuration snippet:

```
            "Tmpfs": {
                "/tmp": "size=1073741824,defaults,noexec"
            },
```

3. Observe that `/tmp` does not have the usual sticky bit set, whereas it did in runc 1.1.7.

### Describe the results you received and expected

`/tmp` sticky bit is not set, but previously it was.

### What version of runc are you using?

1.1.8

### Host OS information

NAME="Ubuntu"
VERSION="20.04.6 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.6 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

### Host kernel information

Linux xxxx 5.15.0-1039-aws #44~20.04.1-Ubuntu SMP Thu Jun 22 12:21:12 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by lifubang**
Fix #3952

When we call `unix.Mount`, if we use file mode bits from the bits with the type `fs.FileMode` directly, it will cause some bits missing.

Please refer: https://github.com/golang/go/blob/master/src/os/file.go#L258-L265

### CODE REVIEW

**Comment by lifubang** [comment]
More details: https://www.gnu.org/software/coreutils/manual/html_node/Mode-Structure.html

File: tests/integration/run.bats [hunk 1]
```diff
@@ -77,6 +77,19 @@ function teardown() {
 	[[ "${lines[0]}" == *'mydomainname'* ]]
 }
 
+@test "runc run with tmpfs" {
+	requires root
+
+	mode=$(stat -c %A /tmp)
```
Comment (cyphar) [inline tests/integration/run.bats:83]: I'm not sure this test is actually testing the fix directly -- you want `rootfs/tmp` to have the same mode even after we mount tmpfs on top, not necessarily the same mode as the host `/tmp`. In addition, I would suggest setting the sticky bit on `rootfs/tmp` explicitly so it's clearer what is being tested. Maybe something like

```suggestion
	chmod 'a=rwx,ug+s,+t' rootfs/tmp # set all bits
	mode=$(stat -c %A rootfs/tmp)
```

Comment (lifubang) [reply tests/integration/run.bats]: Indeed, I was careless just now.
Thanks.

File: libcontainer/mount_linux.go [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package libcontainer
 
 import (
+	"io/fs"
 	"strconv"
+	"syscall"
```
Comment (cyphar) [inline libcontainer/mount_linux.go:6]: (Nit: I prefer using unix for everything but it doesn't really matter.)

File: libcontainer/mount_linux.go [hunk 3]
```diff
@@ -103,3 +105,19 @@ func unmount(target string, flags int) error {
 	}
 	return nil
 }
+
+// SyscallMode returns the syscall-specific mode bits from Go's portable mode bits.
+func SyscallMode(i fs.FileMode) (o uint32) {
```
Comment (kolyshkin) [inline libcontainer/mount_linux.go]: Yeah, golang stdlib does the same conversion in archive/tar and archive/zip.

File: libcontainer/mount_linux.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/mount_linux.go:6]: Same -- I think with Go2 there will be no syscall and we'll have to switch to x/sys/unix anyway, so better do it now.

File: libcontainer/mount_linux.go [hunk 4]
```diff
@@ -106,8 +106,8 @@ func unmount(target string, flags int) error {
 	return nil
 }
 
-// SyscallMode returns the syscall-specific mode bits from Go's portable mode bits.
-func SyscallMode(i fs.FileMode) (o uint32) {
+// syscallMode returns the syscall-specific mode bits from Go's portable mode bits.
+func syscallMode(i fs.FileMode) (o uint32) {
```
Comment (kolyshkin) [inline libcontainer/mount_linux.go]: Can you merge this hunk into the first commit?

File: libcontainer/rootfs_linux.go [hunk 5]
```diff
@@ -462,7 +462,7 @@ func mountToRootfs(c *mountConfig, m mountEntry) error {
 				return err
 			}
 		} else {
-			dt := fmt.Sprintf("mode=%04o", SyscallMode(stat.Mode()))
+			dt := fmt.Sprintf("mode=%04o", syscallMode(stat.Mode()))
```
Comment (kolyshkin) [inline libcontainer/rootfs_linux.go]: same

File: tests/integration/run.bats [hunk 6]
```diff
@@ -77,6 +77,21 @@ function teardown() {
 	[[ "${lines[0]}" == *'mydomainname'* ]]
 }
 
+@test "runc run with tmpfs" {
```
Comment (kolyshkin) [inline tests/integration/run.bats:80]: nit: maybe add a link to issue, e.g.
```bats
# https://github.com/opencontainers/runc/issues/3952
@test "runc run with tmpfs" {
...
```

**Review by kolyshkin — APPROVED** [review]
LGTM, thanks!

**Pull request merged**